compiler-core/src/javascript/typescript.rs

```diff
@@ -0,0 +1,530 @@
+//! This module is responsible for generating Typescript type declaration files.
+//! This code is run during the code generation phase along side the normal
+//! Javascript code emission. Here we walk through the typed AST and translate
+//! the Gleam statements into their Typescript equivalent. Unlike the Javascript
+//! code generation, the Typescript generation only needs to look at the module
+//! statements and not the expressions that may be _inside_ those statements.
+//! This is due to the Typescript declarations only caring about inputs and outputs
+//! rather than _how_ those outputs are generated.
+//!
+//! ## Links
+//! <https://www.typescriptlang.org/>
+//! <https://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html>
+
+use std::{cell::RefCell, rc::Rc, sync::Arc};
+
+use heck::CamelCase;
+use itertools::Itertools;
+
+use crate::{
+    ast::{
+        Statement, TypedArg, TypedConstant, TypedExternalFnArg, TypedModule,
+        TypedRecordConstructor, TypedStatement,
+    },
+    docvec,
+    pretty::{break_, Document, Documentable},
+    type_::{Type, TypeVar},
+};
+
+use super::{concat, import::Imports, line, lines, wrap_args, Output};
+
+const INDENT: isize = 4;
+
+/// The `TypePrinter` contains the logic of how to convert Gleam's type system
+/// into the equivalent Typescript type strings.
+///
+#[derive(Debug)]
+struct TypePrinter<'a> {
+    tracker: Rc<RefCell<UsageTracker>>,
+    current_module: &'a [String],
+}
+
+impl<'a> TypePrinter<'a> {
+    fn new(current_module: &'a [String]) -> Self {
+        Self {
+            current_module,
+            tracker: Rc::new(RefCell::new(UsageTracker::default())),
+        }
+    }
+
+    /// Converts a Gleam type into a Typescript type string
+    ///
+    pub fn print(&self, type_: &Type) -> Document<'static> {
```

Comment:
Could we not store this directly in the TypePrinter and then pass around a mutable reference to that? From what I can see the UsageTracker is not shared elsewhere

```suggestion
#[derive(Debug)]
struct TypePrinter<'a> {
    tracker: UsageTracker,
    current_module: &'a [String],
}

impl<'a> TypePrinter<'a> {
    fn new(current_module: &'a [String]) -> Self {
        Self {
            current_module,
            tracker: UsageTracker::default(),
        }
    }

    /// Converts a Gleam type into a Typescript type string
    ///
    pub fn print(&mut self, type_: &Type) -> Document<'static> {
```

Reply:
Yeah I tried doing that before, but it caused me to have to mark every single function at `mut` as well as every function in the `TypescriptGenerator` as well. But I'll give it another shot for sure!